Add new attachment encrypt/decrypt functions

jagerman opened the pull request:
This defines a new determinstic, streamable attachment encryption/decryption for Session clients to use in the future, built on libsodium's crypto_secretstream API (which is XChaCha20-poly1305-based).

Unlike current Session attachment encryption, the encryption added here uses deterministic (though private, cryptographically secure) nonce and key generated from the user, file contents itself, and type of file, so that the same user uploading an identical attachment results in an identical encrypted copy, thus allowing deduplication in case of things like file server response timeouts or repeated uploads of the same profile picture.

Unlike current attachments, these functions also allow streaming of attachment data (via the added Decryptor) so that, in future clients, an attachment can be decrypted and saved as it arrives, rather than needing to store the whole thing in memory to decrypt it.

This also reimplements the padding added to attachments.  Currently clients use this monstrosity:

    let desiredSize: Int = max(541, min(Int(Network.maxFileSize), Int(floor(pow(1.05, ceil(log(Double(plaintext.count)) / log(1.05)))))))

which is not only difficult to read, but also involves several floating point functions where imprecision (and thus potential metadata leakage about which client created it) can creep in.  The new protocol works vaguely similar: we pad up to fixed sizes that with the size of the attachment.

Moreover there is no way to recover the amount of padding from the attachment itself: rather a Session client has to communicate the URL, key, *and size* to someone they are trying to send to out of band, who then downloads, decrypts the whole thing, then chops off the padding based on what they were told the size was.  With the new scheme, padding is now deterministic and embedded in the pre-encrypted data, and is transparently stripped out when decrypting.

The plan is for Session clients to start understanding (but not using) this new encryption format, and then once sufficient time has passed for clients to have upgraded, start using this instead of the current AES-GCM encryption.

jagerman committed:
Add new attachment encrypt/decrypt functions

This defines a new determinstic, streamable attachment
encryption/decryption for Session clients to use in the future, built on
libsodium's crypto_secretstream API (which is XChaCha20-poly1305-based).

Unlike current Session attachment encryption, the encryption added here
uses deterministic (though private, cryptographically secure) nonce and
key generated from the user, file contents itself, and type of file, so
that the same user uploading an identical attachment results in an
identical encrypted copy, thus allowing deduplication in case of things
like file server response timeouts or repeated uploads of the same
profile picture.

Unlike current attachments, these functions also allow streaming of
attachment data (via the added Decryptor) so that, in future clients, an
attachment can be decrypted and saved as it arrives, rather than needing
to store the whole thing in memory to decrypt it.

This also reimplements the padding added to attachments.  Currently
clients use this monstrosity:

    let desiredSize: Int = max(541, min(Int(Network.maxFileSize), Int(floor(pow(1.05, ceil(log(Double(plaintext.count)) / log(1.05)))))))

which is not only difficult to read, but also involves several floating
point functions where imprecision (and thus potential metadata leakage
about which client created it) can creep in.  The new protocol works
vaguely similar: we pad up to fixed sizes that with the size of the
attachment.

Moreover there is no way to recover the amount of padding from the
attachment itself: rather a Session client has to communicate the URL,
key, *and size* to someone they are trying to send to out of band, who
then downloads, decrypts the whole thing, then chops off the padding
based on what they were told the size was.  With the new scheme, padding
is now deterministic and embedded in the pre-encrypted data, and is
transparently stripped out when decrypting.

The plan is for Session clients to start understanding (but not using)
this new encryption format, and then once sufficient time has passed for
clients to have upgraded, start using this instead of the current
AES-GCM encryption.

mpretty-cyro commented on include/session/attachments.hpp:
```diff
@@ -0,0 +1,208 @@
+///
+/// This class is not recommended if the intention is to build an in-memory buffer from existing
+/// in-memory data: `decrypt()` will be more memory efficient in that case.
+class Decryptor {
```
Do we also need an `Encryptor` to support streaming encryption?

jagerman replied:
The problem with encryption is that we use a hash of the file to generate the key/nonce, and so we really have to be able to go through it twice.

However, we can certainly add a function that encrypts from a file path (it'll just read the file twice -- once to generate the keys, then once to encrypt, but won't have to hold it in memory).

mpretty-cyro replied:
Yea as we chatted about it's not needed for this initial stage but I think it'd be beneficial eventually, because when you want to send a file you would essentially "import" the file into Session first, encrypt it and upload the encrypted content - the client would then use the "imported" version when rendering that message in the conversation history

jagerman replied:
I've pushed an update that adds a bunch of alternative encryption functions, including encryption from plaintext files.

mpretty-cyro commented on include/session/attachments.hpp:
```diff
@@ -0,0 +1,208 @@
+/// decrypting.
+///
+/// Inputs:
+/// - `seed` -- the 32-byte seed of the sender; it is recommended that this be the 32-byte Session
```
Just to confirm - by this you mean the seed used to generate the users `ed25519` key pair?

jagerman replied:
Yeah, but in theory it could be any deterministic 32-byte value if, for example, we needed to generate from some other source.  The seed is perfectly good for any case we have now.

mpretty-cyro commented on include/session/attachments.hpp:
```diff
@@ -0,0 +1,208 @@
+///
+/// Throws std::runtime_error if decryption fails or if writing to the file fails.  Upon exception a
+/// partially written file will be deleted.
+void decrypt(
+        std::span<const std::byte> encrypted,
+        std::span<const std::byte, ENCRYPT_KEY_SIZE> key,
+        const std::filesystem::path& filename);
```
Would it be beneficial to have a version of this which takes the file path of an encrypted file and an output path so clients don't need to use `Decryptor` directly for a file->file decryption?

jagerman replied:
Possibly, yes, but is there a case where that happens that would use it?

mpretty-cyro replied:
The main case I can think of is if you want to save an attachment - the client will ask the user for a destination file path and then we would need to load the encrypted content into memory, the call this function to output the decrypted file to that path (having file -> file would just skip the first step)

For sharing files on iOS or opening document files we would need to decrypt the file and save it to a temporary directory to pass the unencrypted content to the OS, similar to the above it just means we can avoid having to load the encrypted content into memory first

jagerman replied:
There is now a file->file version of `decrypt` (along with a bunch of other alternatives).

mpretty-cyro reviewed (commented):
We also need a C API to support iOS sorry 🥲

jagerman committed:
Add various buffer<->file encryption/decryption; C API

jagerman commented:
I've updated this with a whole bunch of overloads for different combinations of en/decryption to/from vector/buffer/file, and added C API wrappers for the most loved platform of all.

jagerman commented:
Actually added the C header this time.

jagerman committed:
Fix attachment file API bugs

Reading was not working properly, thanks to C++'s horrific std::ifstream
API.  Next time I'm sticking to C APIs for I/O because yuck.
